src/index/api.rs

```diff
@@ -20,137 +27,160 @@ pub(crate) struct CrateOwner {
     pub(crate) name: String,
 }
 
-impl RegistryCrateData {
-    pub(crate) fn get_from_network(pkg: &MetadataPackage) -> Result<Self> {
-        let (release_time, yanked, downloads) = get_release_time_yanked_downloads(pkg)?;
-        let owners = get_owners(pkg)?;
+impl<'a> Api<'a> {
+    pub(super) fn new(api_base: &'a Url) -> Self {
+        Self { api_base }
+    }
+
+    pub(crate) fn get_crate_data(&self, name: &str, version: &str) -> Result<RegistryCrateData> {
+        let (release_time, yanked, downloads) =
+            self.get_release_time_yanked_downloads(name, version)?;
+        let owners = self.get_owners(name)?;
 
-        Ok(Self {
+        Ok(RegistryCrateData {
             release_time,
             yanked,
             downloads,
             owners,
         })
     }
-}
-
-/// Get release_time, yanked and downloads from the registry's API
-fn get_release_time_yanked_downloads(pkg: &MetadataPackage) -> Result<(DateTime<Utc>, bool, i32)> {
-    let url = format!("https://crates.io/api/v1/crates/{}/versions", pkg.name);
-    // FIXME: There is probably better way to do this
-    //        and so many unwraps...
-    let client = Client::new();
-    let mut res = client.get(&url).header(ACCEPT, "application/json").send()?;
-
-    let mut body = String::new();
-    res.read_to_string(&mut body)?;
 
-    let json: Value = serde_json::from_str(&body)?;
-    let versions = json
-        .as_object()
-        .and_then(|o| o.get("versions"))
-        .and_then(|v| v.as_array())
-        .ok_or_else(|| err_msg("Not a JSON object"))?;
-
-    let (mut release_time, mut yanked, mut downloads) = (None, None, None);
-
-    for version in versions {
-        let version = version
+    /// Get release_time, yanked and downloads from the registry's API
+    fn get_release_time_yanked_downloads(
+        &self,
+        name: &str,
+        version: &str,
+    ) -> Result<(DateTime<Utc>, bool, i32)> {
+        let url = {
+            let mut url = self.api_base.clone();
+            url.path_segments_mut()
+                .map_err(|()| err_msg("Invalid API url"))?
+                .extend(&["api", "v1", "crates", name, "versions"]);
```

Comment:
The crates.io API isn't stabilized, see https://rust-lang.github.io/rfcs/2141-alternative-registries.html#registry-index-format-specification. I'm not sure how we handle an error from this function - will it abort the build altogether? I think the right thing to do is default to dummy data, the same way we do if there's no API available.

Reply:
The API is actually publicly documented at <https://doc.rust-lang.org/stable/cargo/reference/registries.html#web-api> with no mention of stability. I don't recall ever seeing any RFCs related to it, so I'm not sure if it's intended to be stable or not.

I also didn't actually check whether the endpoints we use are documented there 😅 I'll take a look at that tomorrow.

Reply:
Actually, I took a look now. This path (`/versions`) isn't specified at all, the other path we use (`/owners`) is.

I think falling back to dummy data on failure makes sense, probably collecting as much as possible and just defaulting the rest.

Reply:
Yeah that seems a good approach to me.